native/src/main/java/io/xlibb/pipe/ResultIterator.java

```diff
@@ -30,26 +31,30 @@
  * Java implementation for the APIs of the stream returned from the pipe.
  */
 public class ResultIterator {
+    protected static final String CLOSING_ERROR = "Closing of a closed pipe is not allowed";
+    private static final String CLOSING_FAILED_ERROR = "Failed to close the stream";
+    private static final String PRODUCE_TO_CLOSED_PIPE_ERROR = "Events cannot be consumed after the stream is closed";
+
     private ResultIterator() {}
 
-    public static Object nextValue(Environment env, BObject streamGenerator) {
+    public static Object nextValue(Environment env, BObject streamGenerator, BTypedesc typeParam) {
         Pipe pipe = (Pipe) streamGenerator.getNativeData(NATIVE_PIPE);
         if (pipe != null) {
             Future future = env.markAsync();
             Callback observer = new Callback(future, pipe.getProducer(), pipe.getTimeKeeper(), pipe.getConsumer());
             long timeout = (long) streamGenerator.getNativeData(TIME_OUT);
-            pipe.asyncConsume(observer, timeout);
+            pipe.asyncConsume(observer, timeout, typeParam.getDescribingType());
             return null;
         }
-        return createError("Events cannot be consumed after the stream is closed");
+        return createError(PRODUCE_TO_CLOSED_PIPE_ERROR);
     }
 
     public static BError close(Environment env, BObject streamGenerator) {
         long timeOut = (long) streamGenerator.getNativeData(TIME_OUT);
         Pipe pipe = ((Pipe) streamGenerator.getNativeData(NATIVE_PIPE));
         if (pipe == null) {
-            BError cause = createError("Closing of a closed pipe is not allowed");
-            return createError("Failed to close the stream", cause);
+            BError cause = createError(CLOSING_ERROR);
+            return createError(CLOSING_FAILED_ERROR, cause);
```

Comment:
What if we have one error which gives the proper meaning and return the error with that ?
```suggestion
            return createError("Attempting to close a closed pipe");
```

Reply:
This `close` method is for the stream. So the first error should be about failing to close the stream. The cause for that should be `Attempting to close a closed pipe`.